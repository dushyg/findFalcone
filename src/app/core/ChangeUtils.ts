import { ISearchAttempt } from "./models/searchAttempt";
import { IPlanet } from "./models/planet";
import { IVehicle } from "./models/vehicle";
import { IFalconAppState } from "./models/falconApp.state";

export class ChangeUtils {
  public static getNextStateAfterChange(
    changedWidgetId: number,
    changedValue: string,
    changedWidgetSearchAttemptGetter: () => ISearchAttempt,
    previousState: IFalconAppState
  ): IFalconAppState {
    const updatedSearchMap = ChangeUtils.getUpdatedSearchMap(
      previousState.searchMap,
      changedWidgetId,
      changedValue,
      changedWidgetSearchAttemptGetter
    );

    // find unsearched planets after this planet change
    const unsearchedPlanets: IPlanet[] = ChangeUtils.getUnsearchedPlanets(
      updatedSearchMap,
      previousState.planetList
    );

    // update the vehicle inventory, taking into account used and freed vehicles
    const updatedVehicleInventory: IVehicle[] = ChangeUtils.getUpdatedVehicleInventory(
      updatedSearchMap,
      previousState.vehicleList
    );

    // calculate the total time taken to search planets with vehicles selected
    const totalTimeTakenForSearch: number = ChangeUtils.getTotalTimeTakenForSearch(
      updatedSearchMap,
      previousState.planetList,
      previousState.vehicleList
    );

    const isReadyToSearch: boolean = ChangeUtils.isItFineToStartSearching(
      updatedSearchMap
    );

    return <IFalconAppState>{
      ...previousState,
      searchMap: updatedSearchMap,
      unsearchedPlanets,
      vehicleInventory: updatedVehicleInventory,
      isReadyToSearch,
      totalTimeTaken: totalTimeTakenForSearch,
    };
  }

  private static getUpdatedSearchMap(
    searchMap: Map<string, ISearchAttempt>,
    changedWidgetId: number,
    changedValue: string,
    changedWidgetSearchAttemptGetter: () => ISearchAttempt
  ): Map<string, ISearchAttempt> {
    let updatedSearchMap = new Map<string, ISearchAttempt>();

    searchMap.forEach((value: ISearchAttempt, key: string) => {
      let updatedSearchAttempt: ISearchAttempt;
      // if the current widget is to the left of changed widget or the changed widget then
      // just make a clone for immutability and continue
      const currentlyLoopedWidgetId = Number(key);
      if (currentlyLoopedWidgetId < changedWidgetId) {
        updatedSearchAttempt = <ISearchAttempt>{ ...value };
      } else if (currentlyLoopedWidgetId == changedWidgetId) {
        updatedSearchAttempt = changedWidgetSearchAttemptGetter();
      } else {
        // else if the current widget is to the right of the changed widget then
        // we will need to reset both the existing planet and vehicle selections if any
        updatedSearchAttempt = <ISearchAttempt>{
          vehicleUsed: null,
          searchedPlanet: null,
        };
      }

      updatedSearchMap.set(key, updatedSearchAttempt);
    });
    return updatedSearchMap;
  }

  private static getUnsearchedPlanets(
    searchMap: Map<string, ISearchAttempt>,
    planetList: IPlanet[]
  ): IPlanet[] {
    const usedPlanetSet = new Set<string>();

    searchMap.forEach((searchAttempt: ISearchAttempt) => {
      if (searchAttempt && searchAttempt.searchedPlanet) {
        usedPlanetSet.add(searchAttempt.searchedPlanet);
      }
    });

    return planetList.filter(
      (planet: IPlanet) => !usedPlanetSet.has(planet.name)
    );
  }

  private static getUpdatedVehicleInventory(
    searchMap: Map<string, ISearchAttempt>,
    vehicleList: IVehicle[]
  ): IVehicle[] {
    // create a map of vehicle name vs count of vehicles used
    const usedVehicleMap = this.getUsedVehicleMap(searchMap);

    // update vehicle list with available vehicle units
    let updatedVehicleList: IVehicle[] = vehicleList.map((vehicle) => {
      const totalNumUnits = vehicle.totalNumUnits;
      return <IVehicle>{
        ...vehicle,
        availNumUnits: totalNumUnits - (usedVehicleMap.get(vehicle.name) || 0),
      };
    });

    return updatedVehicleList;
  }

  private static getUsedVehicleMap(
    searchMap: Map<string, ISearchAttempt>
  ): Map<string, number> {
    const usedVehicleMap = new Map<string, number>();
    for (let index = 1; index < searchMap.size + 1; index++) {
      const key = index.toString();
      const searchAttempt: ISearchAttempt = searchMap.get(key);
      if (searchAttempt) {
        const vehicle: string = searchAttempt.vehicleUsed;
        if (vehicle) {
          let count: number = usedVehicleMap.get(vehicle);
          if (Number(count)) {
            usedVehicleMap.set(vehicle, ++count);
          } else {
            usedVehicleMap.set(vehicle, 1);
          }
        }
      }
    }
    return usedVehicleMap;
  }

  private static getTotalTimeTakenForSearch(
    searchMap: Map<string, ISearchAttempt>,
    allPlanets: IPlanet[],
    allVehicles: IVehicle[]
  ): number {
    let totalTimeTakenForSearch: number = 0;

    const nameToPlanetMap: Map<string, IPlanet> = new Map<string, IPlanet>();
    const nameToVehicleMap: Map<string, IVehicle> = new Map<string, IVehicle>();

    allPlanets.forEach((planet) => {
      nameToPlanetMap.set(planet.name, planet);
    });

    allVehicles.forEach((vehicle) => {
      nameToVehicleMap.set(vehicle.name, vehicle);
    });

    searchMap.forEach((value: ISearchAttempt) => {
      if (!value) {
        return;
      }

      if (!value.searchedPlanet || !value.vehicleUsed) {
        return;
      }

      const planet: IPlanet = nameToPlanetMap.get(value.searchedPlanet);
      const vehicle: IVehicle = nameToVehicleMap.get(value.vehicleUsed);
      const planetDistance = planet.distance;
      const vehicleSpeed = vehicle.speed;
      if (!planetDistance || !vehicleSpeed) {
        return;
      }

      totalTimeTakenForSearch =
        totalTimeTakenForSearch + planetDistance / vehicleSpeed;
    });

    return totalTimeTakenForSearch;
  }

  private static isItFineToStartSearching(
    searchMap: Map<string, ISearchAttempt>
  ): boolean {
    let isReadyForSearch: boolean = true;

    searchMap.forEach((value: ISearchAttempt) => {
      if (!value) {
        isReadyForSearch = false;
        return;
      }

      if (!value.searchedPlanet || !value.vehicleUsed) {
        isReadyForSearch = false;
        return;
      }
    });

    return isReadyForSearch;
  }
}
