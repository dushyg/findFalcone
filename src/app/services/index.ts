export { PlanetsService } from './planets.service';
export { VehiclesService } from './vehicles.service';
export { FalconeTokenService } from './falcone-token.service';
export { FinderFacadeService } from './finder-facade.service';
