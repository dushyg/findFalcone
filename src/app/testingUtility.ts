export const createSpyObj = (
  methodNames: string[]
): { [key: string]: jest.Mock<any> } => {
  const obj: any = {};

  methodNames.forEach((methodName) => {
    obj[methodName] = jest.fn();
  });

  return obj;
};
