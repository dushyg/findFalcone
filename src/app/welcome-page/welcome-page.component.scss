div.welcomePage {
  margin-top: 10vh;
  font-size: 12px;
  h2 {
    text-align: center;
    margin-top: 10px;
  }
  overflow-y: auto;
  div {
    margin-top: 5px;
  }
  div.instructions {
    text-align: left;
    margin-left: 5px;
    margin-right: 5px;
    div.goodLuckText {
      text-align: center;
    }
  }
  div#divStartBtn {
    width: 50px;
    margin: 10px auto;
  }

  @media (min-width: 768px) {
    /*css for medium devices like IPad*/
    font-size: 14px;
    margin-left: 10px;
    margin-right: 10px;
  }

  //   div#divImage {
  //     margin: 0 auto;
  //     img.instructionImg {
  //       max-width: 100%;
  //       height: auto;
  //       background-color: white;
  //     }
  //   }
}
