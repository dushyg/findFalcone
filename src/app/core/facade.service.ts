import { Injectable } from '@angular/core';
import { PlanetsService } from './planets.service';
import { VehiclesService } from './vehicles.service';
import { FalconFinderService } from './falcon-finder.service';
import { Router } from '@angular/router';
import { IFindFalconRequest } from './models/find-falcon-request';
import { IFindFalconResponse } from './models/find-falcon-response';
import { catchError, map, withLatestFrom, tap, distinctUntilChanged } from 'rxjs/operators';
import { Subject, EMPTY, BehaviorSubject, Observable, combineLatest, forkJoin } from 'rxjs';
import PlanetChange from './models/planetChange';
import VehicleChange from './models/vehicleChange';
import PlanetUpdates from './models/planetUpdates';
import { IPlanet } from './models/planet';
import { IVehicle } from './models/vehicle';
import VehicleUpdates from './models/vehicleUpdates';
import { ISearchAttempt } from './models/searchAttempt';

@Injectable({providedIn : 'root'})
export default class FalconeFacade {
            
    constructor(private planetService : PlanetsService,
                private vehicleService : VehiclesService,
                private finderService : FalconFinderService,
                private router : Router){
        
    }

    private finderApiToken : string;       

    private searchMapSubject = new Subject<Map<number, ISearchAttempt>>();
    public searchMap$ : Observable<Map<number, ISearchAttempt>> = this.searchMapSubject.asObservable();

    private readonly MAX_SEARCH_ATTEMPTS_ALLOWED = 4; 
    private maxSearchAttemptAllowedSubject = new BehaviorSubject<number>(this.MAX_SEARCH_ATTEMPTS_ALLOWED);    
    public maxSearchAttemptAllowed$ : Observable<number> = this.maxSearchAttemptAllowedSubject.asObservable();
    public getMaxSearchAttemptsAllowedCount() {
        return this.MAX_SEARCH_ATTEMPTS_ALLOWED;
    }
    private errorMessageSubject = new Subject<string>();
    public errorMessage$ : Observable<string> = this.errorMessageSubject.asObservable();

    private apiVehiclesSubject = new Subject<IVehicle[]>();
    public apiVehicles$ : Observable<IVehicle[]> = this.apiVehiclesSubject.asObservable();

    private vehiclesSubject = new Subject<IVehicle[]>();
    public vehicles$ : Observable<IVehicle[]> = this.vehiclesSubject.asObservable();


    private apiPlanetsSubject = new Subject<IPlanet[]>();
    public apiPlanets$ : Observable<IPlanet[]> = this.apiPlanetsSubject.asObservable();

    private planetsSubject = new Subject<IPlanet[]>();
    public planets$ : Observable<IPlanet[]> = this.planetsSubject.asObservable();


    private totalTimeTakenSubject = new BehaviorSubject<number>(0);    
    public totalTimeTaken$ : Observable<number> = this.totalTimeTakenSubject.asObservable();

    private planetFoundOnSubject = new BehaviorSubject<string>('');
    public planetFoundOn$ : Observable<string> = this.planetFoundOnSubject.asObservable();
    
    private planetChangedSubject = new Subject<PlanetChange>(); 
    private planetChangedAction$ : Observable<PlanetChange> = this.planetChangedSubject.asObservable();
    public planetChanged(planetChange : PlanetChange) {
        this.planetChangedSubject.next(planetChange);
    }
    
    private vehicleChangedSubject = new Subject<VehicleChange>(); 
    private vehicleChangedAction$ : Observable<VehicleChange> = this.vehicleChangedSubject.asObservable();
    public vehicleChanged(vehicleChange : VehicleChange) {
        this.vehicleChangedSubject.next(vehicleChange);
    }
        

    public vehiclesUpdated$ = this.vehicleChangedAction$.pipe( 
        withLatestFrom(this.apiVehicles$, this.apiPlanets$, this.searchMap$),
        map(([vehicleChange, vehicles, planets, searchMap]) => {

            // update list of vehicles with updated available units
            const vehicleUpdates : {updatedVehicles : IVehicle[], lastUpdatedVehicle: IVehicle} 
                        = this.getVehiclesWithUpdatedAvailableUnits(vehicles, vehicleChange, searchMap);
            const updatedVehicles : IVehicle[] = vehicleUpdates.updatedVehicles;
            this.updateVehicleInfo(updatedVehicles);       
            
            // update list of planets available to be searched after this change
            const updatedSearchMap = this.getUpdatedSearchMap(vehicleChange.widgetId, null, vehicleUpdates.lastUpdatedVehicle, searchMap);
            this.setSearchMap(updatedSearchMap);
            const planetsLeftForSearch = this.getPlanetsAvailableForSearch(planets, updatedSearchMap);                                    
            this.updatePlanetListForAvailability(planetsLeftForSearch);
            console.log(`vehiclesUpdated$ -> vehicleChangedAction$ withLatestFrom`,updatedSearchMap);
            return {widgetId : vehicleChange.widgetId, changer : 'vehicleUpdate'};
        })
    )     
    
    public planetsUpdated$ = this.planetChangedAction$
            .pipe(withLatestFrom(this.apiPlanets$, this.apiVehicles$, this.searchMap$),
                  map(([planetChange, planets, vehicles, searchMap]) => {

                        

                        // update list of planets available to be searched after this change
                        const updatedSearchMap = this.getUpdatedSearchMap(planetChange.widgetId, planetChange.newPlanet, null, searchMap);
                        this.setSearchMap(updatedSearchMap);
                        const planetsLeftForSearch = this.getPlanetsAvailableForSearch(planets, updatedSearchMap);                                    
                        this.updatePlanetListForAvailability(planetsLeftForSearch);

                        // update list of vehicles with updated available units
                        const vehicleUpdates : {updatedVehicles : IVehicle[], lastUpdatedVehicle: IVehicle}
                                             = this.getVehiclesWithUpdatedAvailableUnits(vehicles, <VehicleChange>{ widgetId : planetChange.widgetId}, updatedSearchMap);
                        
                        const vehiclesWithUpdatedAvailableUnits: IVehicle[] = vehicleUpdates.updatedVehicles;
                        this.updateVehicleInfo(vehiclesWithUpdatedAvailableUnits);
                        console.log(`planetsUpdated$ -> withLatestFrom`,updatedSearchMap);
                        return {widgetId : planetChange.widgetId, changer : 'planetUpdate'};

                  })
            );

    
    private updateVehicleInfo(updatedVehicles: IVehicle[]): void {

        this.vehiclesSubject.next(updatedVehicles);
    }

    private getUpdatedSearchMap(changedWidgetId: number, changedWidgetPlanet: IPlanet, changedWidgetVehicle : IVehicle, searchMap: Map<number, ISearchAttempt>): Map<number, ISearchAttempt> {
        const newSearchMap = new Map<number, ISearchAttempt>();

        if(!changedWidgetPlanet) {
            changedWidgetPlanet = searchMap.get(changedWidgetId).searchedPlanet;
        }

        //const widgetId = planetChange.widgetId;
        // if there are any existing searchAttempts then
        // include searchAttempts to the left of currently changed widget         
        if(searchMap.size > 0) {            
           searchMap.forEach( (searchAttempt, widgetId) => {
               // this is the searchAttempt for widget to the left of the changed widget and it should be retained
                if(changedWidgetId > widgetId){

                    const unchangedSearchAttemptClone = { ...searchAttempt};
                    newSearchMap.set(widgetId, unchangedSearchAttemptClone);
                }                
           }); 
        }

        // this is the first planet being selected, simply add a new entry to the search map 
        if(changedWidgetPlanet && changedWidgetPlanet.name && changedWidgetPlanet.name !== 'Select'){
            newSearchMap.set(changedWidgetId,
                        <ISearchAttempt>{ 
                            searchedPlanet : {...changedWidgetPlanet},
                            vehicleUsed: changedWidgetVehicle
                            }
                        );
        }

        return newSearchMap;
    }              

    getVehiclesWithUpdatedAvailableUnits(vehicles : IVehicle[], vehicleChange: VehicleChange, searchMap: Map<number, ISearchAttempt>)
                : {updatedVehicles : IVehicle[], lastUpdatedVehicle: IVehicle} 
    {
    
        const updates = {updatedVehicles : null, lastUpdatedVehicle: null};
        let lastUpdatedVehicle : IVehicle;
        const updatedWidgetId = vehicleChange.widgetId;

        // Create a map which will store the updated available units of vehicles against vehicle name
        const vehiclesInUseMap = new Map<string, IVehicle>();
        for(let searchAttemptEntry of searchMap){
            // searchAttemptEntry is an array that has widgetId at index 0 and has serchAttempt at index 1
            const widgetId = searchAttemptEntry[0];
            const vehicleUsed = searchAttemptEntry[1].vehicleUsed; 

            // If current widget is to the left of updated widget
            // the available unit count for the vehicle used in that widget wont change
            if(widgetId < updatedWidgetId){
                if(vehicleUsed) {
                    lastUpdatedVehicle = {...vehicleUsed};
                    vehiclesInUseMap.set(vehicleUsed.name, lastUpdatedVehicle);
                }                
            }
            // We free up one unit from vehicle that was used earlier 
            // and decrease one unit from the newly selected vehicle.
            // If the control reaches here due to planet being updated in which case oldVehicle and newVehicle will be null 
            // then this widget slot will not have any vehicles
            else if(widgetId === updatedWidgetId){                
                    if(vehicleChange.oldVehicle && vehicleChange.oldVehicle.name) {
                        lastUpdatedVehicle = <IVehicle>{ ...vehicleChange.oldVehicle, availNumUnits : vehicleChange.oldVehicle.availNumUnits + 1};
                        vehiclesInUseMap.set(vehicleChange.oldVehicle.name, lastUpdatedVehicle);
                    }
                    if(vehicleChange.newVehicle && vehicleChange.newVehicle.name) {
                        lastUpdatedVehicle = <IVehicle>{ ...vehicleChange.newVehicle, availNumUnits : vehicleChange.newVehicle.availNumUnits - 1};
                        vehiclesInUseMap.set(vehicleChange.newVehicle.name, lastUpdatedVehicle);
                    }                
            }   
            // for all the widgets to the right of the updated widgets we can simply free up all the used units of the vehicles
            // and set availableUnits to totalUnits
            else if(widgetId > updatedWidgetId){
                lastUpdatedVehicle = <IVehicle>{ ...vehicleUsed, availNumUnits : vehicleUsed.totalNumUnits};
                vehiclesInUseMap.set(vehicleUsed.name, lastUpdatedVehicle);
            }
        }

        // Loop over the list of 
        const updatedVehicles : IVehicle[] = vehicles.map( vehicle => {
            const existingVehicle: IVehicle = vehiclesInUseMap.get(vehicle.name);
            
            if(existingVehicle){
                return {...existingVehicle};
            }
            else {
                return {...vehicle};
            }

        });

        
        return {updatedVehicles, lastUpdatedVehicle};
    }    

    private updatePlanetListForAvailability(planets : IPlanet[]) : void {

        this.planetsSubject.next(planets);
    }
    
    private getPlanetsAvailableForSearch(planets: IPlanet[], searchMap : Map<number, ISearchAttempt>)  : IPlanet[] {       
       
        if(searchMap.size > 0) {

            const planetsIncludedInSearch = new Set<string>();

            // create a set of all planets that have already been searched (i.e are present in searchMap)
            for(let searchAttempt of searchMap) {
                //searchAttempt is an array with index 0 being the widget id and index 1 being searchAttempt
                planetsIncludedInSearch.add(searchAttempt[1].searchedPlanet.name)
            }
        
            const planetsAvailableForSearch = planets.filter( planet => {                    
                    // if the planet is part of planetsIncludedInSearch set then filter it out                    
                    return !planetsIncludedInSearch.has(planet.name);
                });
        
            return [...planetsAvailableForSearch];
        }
        else {
            return [...planets];
        }
                        
    }           

    // todo remove, just for testing
    public some$ = combineLatest([this.planets$, this.planetChangedAction$])
    .pipe(
        map(([planets, planetChange]) => {
            
            console.log('combinelatest [this.planets$, this.planetChangedAction$]',planets, planetChange);
            return new PlanetUpdates(planets, planetChange, null);
            
        })
    )
    
    
    public isReadyForSearch$ = this.searchMap$.pipe(
        map(searchMap => {
            // if searchMap contains required entries then return true		
            if(searchMap && searchMap.size === this.MAX_SEARCH_ATTEMPTS_ALLOWED) {
                for(let entry of searchMap){
                    if(!entry[1] ||
                        (entry[1] && (!entry[1].searchedPlanet || !entry[1].vehicleUsed))) {
                        return false;
                    }
                }
                return true;
            }
            return false;
        })
    )        

    public initializeAppData() {

        forkJoin(
          this.vehicleService.getAllVehicles(),
          this.planetService.getAllPlanets(),
          this.finderService.getFalconFinderApiToken()
        )
        .pipe(
            catchError(
                error => {
                    this.setErrorMsg(error);
                    return EMPTY;
                }
            )
        )
        .subscribe( response => {
                      
            const vehicleList : IVehicle[] = response[0];
            const planetList : IPlanet[] = response[1];
            this.finderApiToken = response[2].token;                        

            this.apiVehiclesSubject.next(vehicleList);
            this.vehiclesSubject.next(vehicleList);

            this.apiPlanetsSubject.next(planetList);    
            this.planetsSubject.next(planetList);   
            
            // initialize/reset searchmap, error message and total time taken
            this.setSearchMap(new Map<number, ISearchAttempt>());
            this.errorMessageSubject.next('');
            this.totalTimeTakenSubject.next(0);
        },
          (error) => {
            this.setErrorMsg(error);
          }
        );

        this.searchMap$.pipe(
            tap(searchMap => {
                if(searchMap.size > 0){
                    let totalTimeTaken = 0;
                    for(let entry of searchMap){                        
                        let searchAttempt = entry[1];
                        if(searchAttempt.searchedPlanet && searchAttempt.searchedPlanet.name &&
                           searchAttempt.vehicleUsed && searchAttempt.vehicleUsed.name && searchAttempt.vehicleUsed.speed){

                                totalTimeTaken +=  (searchAttempt.searchedPlanet.distance / searchAttempt.vehicleUsed.speed);
                           }
                    }
                    this.totalTimeTakenSubject.next(totalTimeTaken);
                }
            }),
            distinctUntilChanged()
        ).subscribe( totalTime => {});

    }   
    
    private setSearchMap(searchMap: Map<number, ISearchAttempt>) : void {
        this.searchMapSubject.next(searchMap);
    }
        
    public findFalcon(request : IFindFalconRequest) {

        request.token = this.finderApiToken; 
        this.finderService.findFalcon(request)
           .subscribe( (response: IFindFalconResponse) => {
             let errorMsg = null;  
             if(response) {
   
               if(response.error) {
                 errorMsg = response.error;
                 this.setErrorMsg(errorMsg);
               }
               else if(response.status) {
   
                 if(response.status.trim().toLowerCase() === "success") {
   
                   if(response.planetName) {
                     this.planetFoundOnSubject.next(response.planetName);
                   }
                   else {
                     errorMsg = "Search api returned empty planet name";
                     this.setErrorMsg(errorMsg);
                   }
                 }
                 else if(response.status.trim().toLocaleLowerCase() === "false") {
                   errorMsg = "Failure! You were unable to find Falcone. Better luck next time.";
                   this.setErrorMsg(errorMsg);
                 }
                 else {
                   errorMsg = "Search api did not return a response status value."
                   this.setErrorMsg(errorMsg);
                 }
               }
             }
             else {
               errorMsg = "Search api returned invalid response."
               this.setErrorMsg(errorMsg);
             }
           },
           (error) => {
            this.setErrorMsg(error);
           }
           );
     }    

    private setErrorMsg(error) {
        this.errorMessageSubject.next(error);
    }

     resetApp(){
        this.router.navigate(['/finderboard/reset']);
      }      
}