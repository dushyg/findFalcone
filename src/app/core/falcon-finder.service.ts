import { Injectable } from "@angular/core";
import { HttpClient, HttpHeaders } from "@angular/common/http";
import { catchError, map } from "rxjs/operators";
import { handleError } from "./handleError";
import { Observable } from "rxjs";
import { IFindFalconeRequest } from "./models/findFalconeRequest";
import { IFindFalconeResponse } from "./models/findFalconeResponse";

@Injectable({
  providedIn: "root",
})
export class FalconFinderService {
  constructor(private http: HttpClient) {}

  private readonly falconFinderApiUrl =
    "https://findfalcone.herokuapp.com/find";

  public findFalcon(
    request: IFindFalconeRequest
  ): Observable<IFindFalconeResponse> {
    const httpOptions = {
      headers: new HttpHeaders({
        Accept: "application/json",
        "Content-Type": "application/json",
      }),
    };
    return this.http
      .post<IFindFalconeResponse>(this.falconFinderApiUrl, request, httpOptions)
      .pipe(
        map((response: any) => {
          return <IFindFalconeResponse>{
            planetName: response.planet_name,
            status: response.status,
            error: response.error,
          };
        }, catchError(handleError))
      );
  }
}
