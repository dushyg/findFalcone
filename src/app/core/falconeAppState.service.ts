import { Injectable, OnInit } from '@angular/core';
import { ISearchAttempt } from './models/searchAttempt';
import { IPlanet } from './models/planet';
import { IVehicle } from './models/vehicle';
import { IFalconAppState } from './models/falconApp.state';
import { CoreModule } from './core.module';
import { BehaviorSubject, Observable } from 'rxjs';

@Injectable({providedIn : CoreModule})
export class FalconeAppStateService
{
    
    private _state : IFalconAppState = {
        errorMsg : "",
        isLoading : false,
        maxSearchAttemptAllowed : 4,
        planetList : [],
        searchMap : new Map<string, ISearchAttempt>(),
        totalTimeTaken : 0,
        vehicleList : [],
        planetFoundOn : null,
        isReadyToSearch : false
      };            
      
    private store = new BehaviorSubject<IFalconAppState>(this._state);
      // private methods

  public getStore() : Observable<IFalconAppState> {

    return this.store.asObservable();
  }
      
  public getAppState() : IFalconAppState {
      return this._state;
  }
    
  public setAppState(state : IFalconAppState) {

    // update the _state variable to reflect latest state
    this._state = state;

    this.store.next(this._state);
    
  }
  
  public updatePlanetFoundOn(planetFoundOn : string) : void {
    
    this.setAppState({ ...this.getAppState(), planetFoundOn});    
  }
  
  public setMaxSearchAttemptAllowed(maxSearchAttemptAllowed : number) : void {

    this.setAppState({...this.getAppState(), maxSearchAttemptAllowed});
  }
    
  public updateError(error : any) : void { 
    console.log('setError', error);
    const errorMsg = error;
    this.setAppState({ ...this.getAppState(), errorMsg});
  }      
  
  public getReadyWidgetState(searchAttemptMap : Map<string, ISearchAttempt>) : { readyWidgetCount : number, totalTimeTaken : number} {

    let readyWidgetCount = 0;
    let totalTimeTaken = 0;

      for(let searchAttempt of searchAttemptMap.values()) {

          if(searchAttempt.vehicleUsed && searchAttempt.searchedPlanet) {
            readyWidgetCount++;
            totalTimeTaken += (searchAttempt.searchedPlanet.distance / searchAttempt.vehicleUsed.speed);
          }          
      }

      return {
        readyWidgetCount ,
        totalTimeTaken
      };
  }

  // selector methods

  public getVehicleList() : IVehicle[] {

    return this.getAppState().vehicleList;
  }

  public getPlanetList() : IPlanet[] {

    return this.getAppState().planetList;
  }

  public getVehicleByName(vehicleName : string) : IVehicle {

    if(vehicleName) {
      return this.getVehicleList().find( v => v.name === vehicleName);
    }
    return null;
  }

  public getPlanetByName(planetName : string) : IPlanet {

    if(planetName) {
      return this.getPlanetList().find( p => p.name === planetName);
    }
    return null;
  }

  public getSearchMap() : Map<string, ISearchAttempt> {

    return this.getAppState().searchMap;
  }

  public getMaxSearchAttemptAllowed()  : number {

    return this.getAppState().maxSearchAttemptAllowed;
  }
  
}