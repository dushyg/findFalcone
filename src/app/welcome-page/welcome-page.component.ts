import { Component, ChangeDetectionStrategy } from '@angular/core';
import { Router } from '@angular/router';
import { FinderFacadeService } from '../services';

@Component({
  selector: 'app-welcome-page',
  templateUrl: './welcome-page.component.html',
  styleUrls: ['./welcome-page.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class WelcomePageComponent {
  constructor(
    private router: Router,
    private finderFacadeService: FinderFacadeService
  ) {}

  public start() {
    this.finderFacadeService.setLoadingFlag(true);
    this.router.navigate(['/finderboard']);
  }
}
