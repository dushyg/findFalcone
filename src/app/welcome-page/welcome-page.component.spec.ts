import { ComponentFixture, TestBed } from '@angular/core/testing';
import { WelcomePageComponent } from './welcome-page.component';
import { Router } from '@angular/router';
import { createSpyObj } from '../testingUtility';
import { FinderFacadeService } from '../services';

describe('Welcome Page Component', () => {
  let fixture: ComponentFixture<WelcomePageComponent>;
  let component: WelcomePageComponent;
  let routerServiceMock: { [key: string]: jest.Mock<any, any> };
  let finderFacadeServiceMock;

  beforeEach(() => {
    routerServiceMock = createSpyObj(['navigate']);
    finderFacadeServiceMock = {
      setLoadingFlag: (loading) => {},
    };
    TestBed.configureTestingModule({
      declarations: [WelcomePageComponent],
      providers: [
        { provide: Router, useValue: routerServiceMock },
        { provide: FinderFacadeService, useValue: finderFacadeServiceMock },
      ],
    });
    fixture = TestBed.createComponent(WelcomePageComponent);
    component = fixture.componentInstance;
  });

  it('Is created', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
    (expect(fixture) as any).toMatchSnapshot();
  });

  it('navigates to finderboard route when start button is clicked', () => {
    fixture.detectChanges();
    const loadingSpy = jest.spyOn(finderFacadeServiceMock, 'setLoadingFlag');
    const routerSpy = jest.spyOn(routerServiceMock, 'navigate');
    const btnStart: HTMLButtonElement = fixture.nativeElement.querySelector(
      '.btnStart'
    );

    expect(btnStart).toBeTruthy();

    btnStart.click();

    expect(loadingSpy).toHaveBeenCalledWith(true);
    expect(routerSpy).toHaveBeenCalledWith(['/finderboard']);
  });
});
