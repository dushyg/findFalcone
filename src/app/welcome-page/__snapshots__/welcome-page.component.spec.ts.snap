// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome Page Component Is created 1`] = `
<app-welcome-page
  __ngContext__={[Function LRootView]}
  finderFacadeService={[Function Object]}
  router={[Function Object]}
>
  <div
    class="welcomePage"
  >
    <h2>
      Instructions
    </h2>
    <div
      class="instructions"
    >
       After the recent Falicornian war, King Shan has exiled Queen Al Falcone for 15 years. However, if he finds her before the 15 years are up, she has to go into exile for another 15 years! King Shan has received intelligence that Al Falcone is hiding in one of six neighbouring planets. To solve this problem this application helps you to choose the planets to search, and the vehicles to use in Finding Falcone. 
      <br />
      <br />
       You need to search across 4 planets from a list and use one of the available vehicles to search a planet. Every vehicle has its a limit for distance it can cover. So it cant be used for a planet that is at a distance greater than the max it can travel. Once a planet is selected it is removed from list of available planets. Once a vehicle is selected its available count is reduced by 1 in Vehicle Inventory. 
      <br />
      <br />
       To honor above constraints, if a planet or vehicle is set / changed, all planets and vehicles to the right are reset. Planet searched & vehicle used for search are grouped into a destination widget. There is a typeahead control per widget that allows user to easily search planets from a list. As soon as user finishes setting up all planets to be searched and vehicles used for the search, the 'Find' button is enabled to initiate the search. 
      <br />
      <br />
      <div
        class="goodLuckText"
      >
        Good luck Finding Falcone!
      </div>
    </div>
    <div
      id="divStartBtn"
    >
      <input
        class="btnStart"
        type="button"
        value="Start"
      />
    </div>
  </div>
</app-welcome-page>
`;
